model/integration_action_test.go

```diff
@@ -109,3 +109,52 @@ func TestPostActionIntegrationResponseFromJsonError(t *testing.T) {
 	ro := PostActionIntegrationResponseFromJson(strings.NewReader(""))
 	assert.Nil(t, ro)
 }
+
+func TestSubmitDialogRequestToJson(t *testing.T) {
+	t.Run("all fine", func(t *testing.T) {
+		request := SubmitDialogRequest{
+			URL:        "http://example.org",
+			CallbackId: NewId(),
+			State:      "some state",
+			UserId:     NewId(),
+			ChannelId:  NewId(),
+			TeamId:     NewId(),
+			Submission: map[string]interface{}{
+				"text":  "some text",
+				"float": 1.2,
+				"bool":  true,
+			},
+			Cancelled: true,
+		}
+		jsonRequest := request.ToJson()
+		r := SubmitDialogRequestFromJson(bytes.NewReader(jsonRequest))
```

Comment:
nit: 0/5, but I find it useful not to have a circular dependency of correctness in a unit test, but to test the output explicitly. In this case (and below), I might recommend just writing out the expected JSON encoding, and doing a simple string comparison.

It's not really a problem in this case, but you could imagine a `assert.Equal(t, x, to(from(x)))` that would pass even if `to` and `from` were the identity functions.

Reply:
I agree that having a circular dependency is not ideal. But since this is a small method and the code has been used before, I think it not worth it to manually write JSON output.

Reply:
Agreed -- I just ran into this myself, and also hated the idea of writing out JSON, so instead I repeated the test twice with two instances of the model in question. </musings>